ibridges/search.py

```diff
@@ -47,90 +60,146 @@ def search_data(
+            _key_filter(key, *queries)
+        else:
+            for sub_key in key:
+                _key_filter(sub_key, *queries)
+
+    if value is not ...:
+        if isinstance(value, str):
+            _value_filter(value, *queries)
+        else:
+            for sub_val in value:
+                _value_filter(sub_val, *queries)
+
+    if units is not ...:
+        if isinstance(units, str):
+            _units_filter(units, *queries)
+        else:
+            for sub_units in units:
+                _units_filter(sub_units, *queries)
 
     results = []
-    if checksum:
+    if checksum is not None:
+        if not checksum.startswith("sha2:"):
+            checksum = f"sha2:{checksum}"
         data_query = data_query.filter(icat.LIKE(icat.DATA_CHECKSUM, checksum))
-        data_name_query = data_name_query.filter(icat.LIKE(icat.DATA_CHECKSUM, checksum))
+        data_name_query.filter(icat.LIKE(icat.DATA_CHECKSUM, checksum))
     else:
-        # gather collection results
-        coll_res = list(coll_query.get_results())
-        if len(coll_res) > 0:
-            results.extend(coll_res)
+        # Gather collection results (not needed if checksum is used).
+        results = list(coll_query.get_results())
+
 
     # gather results, data_query and data_name_query can contain the same results
     results.extend([
         dict(s) for s in set(frozenset(d.items())
                 for d in list(data_query) + list(data_name_query))
     ])
-
     for item in results:
         if isinstance(item, dict):
             new_keys = [k.icat_key for k in item.keys()]
             for n_key, o_key in zip(new_keys, item.keys()):
                 item[n_key] = item.pop(o_key)
 
-    return results
-
-def _path_params(path_param):
-    """Parse the path parameter and return, path, name and parent."""
-    if path_param:
-        path = str(path_param)
-        if len(path.rsplit("/", maxsplit=1)) > 1:
-            parent = path.rsplit("/", maxsplit=1)[0]
-            name = path.rsplit("/", maxsplit=1)[1]
+    # Convert the results to IrodsPath objects.
+    ipath_results = []
+    for res in results:
+        if "DATA_NAME" in res:
+            ipath_results.append(IrodsPath(session, res["COLL_NAME"], res["DATA_NAME"]))
         else:
-            name = path
-            parent = "%"
-        return path, name, parent
-    return None
+            ipath_results.append(IrodsPath(session, res["COLL_NAME"]))
+    return ipath_results
+
+
+def _prefix_wildcard(pattern):
+    if pattern.startswith("%"):
+        return pattern
+    return f"%/{pattern}"
+
+def _path_filter(root_path, path_pattern, coll_query, data_query, data_name_query):
+    coll_query.filter(icat.LIKE(icat.COLL_NAME, _prefix_wildcard(path_pattern)))
+    split_pat = path_pattern.rsplit("/", maxsplit=1)
+    data_query.filter(icat.LIKE(icat.DATA_NAME, split_pat[-1]))
+    data_name_query.filter(icat.LIKE(icat.DATA_NAME, split_pat[-1]))
+
+    if len(split_pat) == 2:
```

Comment:
I think this will return more results than the search defined?

If I am looking for `/my_coll/%/books`
then `_prefix_wildcard(split_pat[0])` will not only return  all `books` collections under `my_coll`, but everything?! Do I understand the code correctly?

Reply:
Or is there then another filter on the query that specifies the data name or the coll name somewhere up in the code?

Reply:
For your query I would use `search_data(session, IrodsPath(session, "/my_coll"), path_pattern="books")`, so you wouldn't need any wildcards for that anymore. Otherwise `search_data(session, IrodsPath(session, "/"), path_pattern="/my_coll/%/books")` should also work I think.

Reply:
But indeed good to test these edge cases.